LiS/man/man4/nuls.4.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) $Id: nuls.4.man,v 0.9.2.3 2008-04-29 08:33:19 brian Exp $
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date: 2008-04-29 08:33:19 $ by $Author: brian $
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log: nuls.4.man,v $
.\" Revision 0.9.2.3  2008-04-29 08:33:19  brian
.\" - update headers for Affero release
.\"
.\" Revision 0.9.2.2  2006/04/05 09:42:35  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\"
.\"
.TH NULS 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B nuls
\- null \fISTREAMS\fR device
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <unistd.h>
.br
.B #include <stropts.h>
.HP 12
.BI "int " fd " = open(" /dev/nuls ", " oflag );
.\"
.\"
.SH DESCRIPTION
.PP
.I fd
is a file descriptor for a null stream opened using the
.BR open (2)
system call on one of the following devices:
.TP
.RB \*(lq /dev/nuls \*(rq
This device is accessible to all users.
.PP
The
.B nuls
device is characterized by the following behaviour:
.IP \(em \w'\(em\(em'u
All data written to  the
.B nuls
device with
.BR write (2),
.BR writev (2),
.BR putmsg (2)
and
.BR putpmsg (2s)
is discarded.
.IP \(em
All attempts to read the
.B nuls
device with
.BR read (2),
.BR readv (2),
.BR getmsg (2)
or
.BR getpmsg (2s)
will return zero length data.
.PP
In contrast to the Linux
.B /dev/null
and
.B /dev/zero
(see
.BR null (4),
.BR zero (4)),
.B nuls
is a
.B STREAMS
null device with the following additional characteristics:
.IP \(em \w'\(em\(em'u
.I STREAMS
modules can be pushed and popped from the device.
.IP \(em
The device response to all
.I STREAMS
I/O controls described in
.BR streamio (7).
.IP \(em
The device can be linked under a multiplexing driver.
.\"
.\"
.SH IOCTLS
.PP
.B nuls
does not support any I/O controls of its own and all I/O controls delivered to
the driver will return an
.IR errno (3)
of
.RB [ EOPNOTSUPP ].
.\"
.\"
.SH USAGE
.PP
The
.B nuls
device can be opened when no other suitable device exists and modules may be
pushed on the
.B nuls
device.
.PP
Pushing a
.I STREAMS
module that responds to administrative I/O controls on the
.B nuls
device is a method of providing administrative drivers without consuming major device
numbers.
.\"
.\"
.SH NOTICES
.PP
The
.B nuls
device is used on systems that provide the older
.IR "SVR 3"
.[
svr3
.]
style
.B /dev/null
or
.BR /dev/zero .
On those systems, the
.B /dev/null
and
.B /dev/zero
devices are not
.I STREAMS
based and will not respond to
.I STREAMS
I/O controls.
.PP
The
.B nuls
device is used with the
.BR sc (4)
module as on
.IR AIX \(rg,
.IR HP-UX \(rg,
and
.IR OSF/1 \(rg
systems in support of the
.BR autopush (8),
.BR scls (8),
.BR strinfo (8),
.BR strload (8),
and
.BR strsetup (8)
commands.
.PP
Under
.B Linux Fast-STREAMS
.[
lfs
.]
the
.B nuls
device is also used with the
.BR sockmod (4)
module in support of
.BR libsocket ,
and by
.BR libstreams ,
as an anonymous stream for non-file-descriptor related system call emulation
for
.BR fdetach (3).
.\"
.\"
.SH IMPLEMENTATION
.PP
.B nuls
is implemented as a regular
.I STREAMS
pseudo-device driver.
.\"
.\"
.\".SH EXAMPLES
.\".PP
.\"
.\"
.SH DEVICES
.PP
.TP
.B /dev/nuls
The external filesystem null
.I STREAMS
device.
.TP
.BR /dev/streams/nuls ", " /devices/nuls
The
.BR specfs (5)
shadow special filesystem null
.I STREAMS
device.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR null (4),
.BR zero (4),
.BR echo (4),
.\"
.\"
.SH BUGS
.PP
.B nuls
has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
The
.B nuls
device is compatible with
.IR "SVR 4.2" ,
.IR AIX \(rg,
.IR HP-UX \(rg,
and
.IR OSF/1 \(rg
with the following portability considerations:
.IP \(em \w'\(em\(em'u
.B /dev/nuls
can be used instead of
.B /dev/null
or
.B /dev/zero
for any purpose for which those devices are used.
.\"
.\"
.SH CONFORMANCE
.PP
.IR "SVR 4.2" ,
.[
svr42cr
.]
and
.IR AIX \(rg,
.IR HP-UX \(rg,
and
.IR OSF/1 \(rg
documentation.
.[
aixdoc
.]
.[
hpuxdoc
.]
.[
osfdoc
.]
Conformance is verified using the
.BR test-nuls (8)
verification test suite.
.\"
.\"
.SH HISTORY
.PP
The
.B nuls
device first appeared in
.I SVR 4.2
.[
svr42
.]
implementations for
.IR AIX \(rg,
.IR HP-UX \(rg,
and
.IR OSF/1 \(rg.

.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 Master Package -- history of user visible changes.  2008-05-03
$Id: NEWS,v 0.9.2.53 2008-05-03 21:22:25 brian Exp $
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2000  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
OpenSS7 Master Package package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Documented `lib32gcc1' problem on Ubuntu.

   - Noted problem running under XEN kernels.  XEN kernels are not yet
     supported.

   - License upgrade to AGPL v3.

   - Cross-compile fixes (strap out AC_FUNC_REALLOC macro when
     cross-compiling).

   - Ability to strap out major documentation build and installation
     primarily for embedded targets.

   - Improvements to common build process for embedded and
     cross-compile targets.

   - Support for flex 2.5.33  in maintainer mode.

   - Added the following sub-packages to the master build (and relase):

         - `strx25-0.9.2.1'


_This is a public stable production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
RPM- and DPKG-based distributions, and can be used on production
kernels without patching or recompiling the kernel.

This package is publicly released under the `GNU Affero General Public
License Version 3'.  The release is available as an `autoconf' tarball,
SRPM, DSC, and set of binary RPMs and DEBs.  See the downloads page
(http://www.openss7.org/download.html) for the `autoconf' tarballs,
SRPMs and DSCs.  For tarballs, SRPMs, DSCs and binary RPMs and DEBs,
see the openss7 package page (http://www.openss7.org/openss7_pkg.html).

See `http://www.openss7.org/codefiles/openss7-0.9.2.G/ChangeLog' and
`http://www.openss7.org/codefiles/openss7-0.9.2.G/NEWS' in the release
for more information.  Also, see the `openss7.pdf' manual in the
release (also in html `http://www.openss7.org/openss7_manual.html').

 Major changes for release openss7-0.9.2.F
------------------------------------------

This is the fifth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Support build on openSUSE 10.2.

   - Support build on Fedora 7 and 2.6.21 kernel.

   - Support build on CentOS 5.0 (RHEL5).

   - Support build on Ubuntu 7.04.

   - Updated to gettext 0.16.1.

   - Changes to support build on 2.6.20-1.2307.fc5 and
     2.6.20-1.2933.fc6 kernel.

   - Supports build on Fedora Core 6.

   - Support for recent distributions and tool chains.


 Major changes for release openss7-0.9.2.E
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

It is unlikely that the `OpenSS7 Master Package' will be released as
frequently as before.  Sub-packages will be released more often.  To
rebuild the master package with a new sub-package release, simply
replace the directory to which the sub-package belongs with the
unpacked sub-package release and then rebuild the master package.  This
release provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Addition of the `osr61' sub-package that contains Dialogic(R) Open
     System Release 6.1 version 239 GPL drivers.

   - A few minor corrections to the common build process.

   - Support for autoconf 2.61, automake 1.10 and gettext 0.16.

   - Support for Ubuntu 6.10 distribution and bug fixes for i386
     kernels.

   - The package now looks for subpackages with a version number as
     unpacked by separate tarball.


 Major changes for release openss7-0.9.2.D
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  The
sub-packages have been reorganized for this release.  See `README' in
the release for a sub-package listing.  Aside from sub-package
reorganization, the major difference from previous release is that this
release no longer contains `LiS'.  Too many of the sub-packages will
not even build against `LiS' because of its Stream head deficiencies.


LiS/drivers/str/nuls.c
/*****************************************************************************

 @(#) $RCSfile: nuls.c,v $ $Name:  $($Revision: 0.9.2.16 $) $Date: 2008-04-29 08:33:11 $

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date: 2008-04-29 08:33:11 $ by $Author: brian $

 -----------------------------------------------------------------------------

 $Log: nuls.c,v $
 Revision 0.9.2.16  2008-04-29 08:33:11  brian
 - update headers for Affero release

 Revision 0.9.2.15  2007/08/15 04:57:58  brian
 - GPLv3 updates

 Revision 0.9.2.14  2007/08/14 10:46:56  brian
 - GPLv3 header update

 *****************************************************************************/

#ident "@(#) $RCSfile: nuls.c,v $ $Name:  $($Revision: 0.9.2.16 $) $Date: 2008-04-29 08:33:11 $"

static char const ident[] =
    "$RCSfile: nuls.c,v $ $Name:  $($Revision: 0.9.2.16 $) $Date: 2008-04-29 08:33:11 $";

#include <linux/autoconf.h>
#include <linux/version.h>
#include <linux/module.h>
#include <linux/init.h>

#include <sys/kmem.h>
#include <sys/stream.h>
#include <sys/strconf.h>
#ifdef LFS
#include <sys/strsubr.h>
#endif
#include <sys/ddi.h>

#ifdef LFS
#include "sys/config.h"
#endif

#define NULS_DESCRIP	"UNIX SYSTEM V RELEASE 4.2 FAST STREAMS FOR LINUX"
#define NULS_COPYRIGHT	"Copyright (c) 1997-2008 OpenSS7 Corporation.  All Rights Reserved."
#define NULS_REVISION	"LfS $RCSfile: nuls.c,v $ $Name:  $($Revision: 0.9.2.16 $) $Date: 2008-04-29 08:33:11 $"
#define NULS_DEVICE	"SVR 4.2 STREAMS Null Stream (NULS) Device"
#define NULS_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define NULS_LICENSE	"GPL"
#define NULS_BANNER	NULS_DESCRIP	"\n" \
			NULS_COPYRIGHT	"\n" \
			NULS_REVISION	"\n" \
			NULS_DEVICE	"\n" \
			NULS_CONTACT	"\n"
#define NULS_SPLASH	NULS_DEVICE	" - " \
			NULS_REVISION	"\n"

#if defined LIS && defined MODULE
#define CONFIG_STREAMS_NULS_MODULE MODULE
#endif

#ifdef CONFIG_STREAMS_NULS_MODULE
MODULE_AUTHOR(NULS_CONTACT);
MODULE_DESCRIPTION(NULS_DESCRIP);
MODULE_SUPPORTED_DEVICE(NULS_DEVICE);
MODULE_LICENSE(NULS_LICENSE);
#if defined MODULE_ALIAS
MODULE_ALIAS("streams-nuls");
#endif
#endif

#ifndef CONFIG_STREAMS_NULS_NAME
#ifdef LIS
#define CONFIG_STREAMS_NULS_NAME NULS__DRV_NAME
#endif
#ifdef LFS
#define CONFIG_STREAMS_NULS_NAME "nuls"
#endif
#endif
#ifndef CONFIG_STREAMS_NULS_MODID
#ifdef LIS
#define CONFIG_STREAMS_NULS_MODID NULS__ID
#endif
#ifdef LFS
#error CONFIG_STREAMS_NULS_MODID must be defined.
#endif
#endif
#ifndef CONFIG_STREAMS_NULS_MAJOR
#ifdef LIS
#define CONFIG_STREAMS_NULS_MAJOR NULS__CMAJOR_0
#endif
#ifdef LFS
#error CONFIG_STREAMS_NULS_MAJOR must be defined.
#endif
#endif

modID_t modid = CONFIG_STREAMS_NULS_MODID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else
module_param(modid, ushort, 0444);
#endif
MODULE_PARM_DESC(modid, "Module id number for NULS driver. (0 for auto allocation)");

#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-modid-" __stringify(CONFIG_STREAMS_NULS_MODID));
MODULE_ALIAS("streams-driver-nuls");
#endif

major_t major = CONFIG_STREAMS_NULS_MAJOR;

#ifndef module_param
MODULE_PARM(major, "h");
#else
module_param(major, uint, 0444);
#endif
MODULE_PARM_DESC(major, "Major device number for NULS driver. (0 for auto allocation)");

#ifdef MODULE_ALIAS
MODULE_ALIAS("char-major-" __stringify(CONFIG_STREAMS_NULS_MAJOR) "-*");
MODULE_ALIAS("/dev/nuls");
#ifdef LFS
MODULE_ALIAS("streams-major-" __stringify(CONFIG_STREAMS_NULS_MAJOR));
MODULE_ALIAS("/dev/streams/nuls");
MODULE_ALIAS("/dev/streams/nuls/*");
MODULE_ALIAS("/dev/streams/clone/nuls");
#endif
#endif

#ifdef LIS
#define STRMINPSZ   0
#define STRMAXPSZ   4096
#define STRHIGH	    5120
#define STRLOW	    1024
#endif

static struct module_info nuls_minfo = {
	.mi_idnum = CONFIG_STREAMS_NULS_MODID,
	.mi_idname = CONFIG_STREAMS_NULS_NAME,
	.mi_minpsz = STRMINPSZ,
	.mi_maxpsz = STRMAXPSZ,
	.mi_hiwat = STRHIGH,
	.mi_lowat = STRLOW,
};

static struct module_stat nuls_rstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));
static struct module_stat nuls_wstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));

#ifdef LIS
#define _trace() while (0) { }
#define _ptrace(__x) while (0) { }
#define _printd(__x) while (0) { }
#define pswerr(__x) while (0) { }

union ioctypes {
	struct iocblk iocblk;
	struct copyreq copyreq;
	struct copyresp copyresp;
};

#define streamscall _RP

#endif

static streamscall int
nuls_put(queue_t *q, mblk_t *mp)
{
	int err = 0;

	_trace();
	switch (mp->b_datap->db_type) {
	case M_FLUSH:
		if (mp->b_rptr[0] & FLUSHW) {
			if (mp->b_rptr[0] & FLUSHBAND)
				flushband(q, mp->b_rptr[1], FLUSHDATA);
			else
				flushq(q, FLUSHDATA);
			mp->b_rptr[0] &= ~FLUSHW;
		}
		if (mp->b_rptr[0] & FLUSHR) {
			queue_t *rq = RD(q);

			if (mp->b_rptr[0] & FLUSHBAND)
				flushband(rq, mp->b_rptr[1], FLUSHDATA);
			else
				flushq(rq, FLUSHDATA);
			qreply(q, mp);
			return (0);
		}
		break;
	case M_IOCTL:
		err = -EINVAL;
		goto nak;
	case M_IOCDATA:
		err = -EINVAL;
		goto nak;
	case M_READ:
		mp->b_wptr = mp->b_rptr;
		mp->b_datap->db_type = M_DATA;
		mp->b_flag |= MSGDELIM;
		qreply(q, mp);
		return (0);
	default:
		freemsg(mp);
		return (0);
	}
      nak:
	{
		union ioctypes *ioc;

		mp->b_datap->db_type = M_IOCNAK;
		ioc = (typeof(ioc)) mp->b_rptr;
		ioc->iocblk.ioc_count = 0;
		ioc->iocblk.ioc_rval = -1;
		ioc->iocblk.ioc_error = -err;
		qreply(q, mp);
		return (0);
	}
}

typedef struct nuls {
	struct nuls *next;
	struct nuls **prev;
	dev_t dev;
} nuls_t;

static spinlock_t nuls_lock = SPIN_LOCK_UNLOCKED;
static struct nuls *nuls_list = NULL;

/* 
 *  -------------------------------------------------------------------------
 *
 *  OPEN and CLOSE
 *
 *  -------------------------------------------------------------------------
 */
static streamscall int
nuls_open(queue_t *q, dev_t *devp, int oflag, int sflag, cred_t *crp)
{
	struct nuls *p, **pp = &nuls_list;
	major_t cmajor = getmajor(*devp);
	minor_t cminor = getminor(*devp);

	_ptrace(("%s: opening major %hu, minor %hu, sflag %d\n", __FUNCTION__, cmajor, cminor,
		 sflag));
	if (q->q_ptr != NULL) {
		_printd(("%s: stream is already open\n", __FUNCTION__));
		return (0);	/* already open */
	}
	if (sflag == MODOPEN || WR(q)->q_next) {
		_printd(("%s: cannot open as module\n", __FUNCTION__));
		return (ENXIO);	/* can't open as module */
	}
	if (!(p = kmem_alloc(sizeof(*p), KM_NOSLEEP))) {	/* we could sleep */
		_printd(("%s: could not allocate private structure\n", __FUNCTION__));
		return (ENOMEM);	/* no memory */
	}
	bzero(p, sizeof(*p));
	switch (sflag) {
	case CLONEOPEN:
		_printd(("%s: clone open\n", __FUNCTION__));
		if (cminor < 1)
			cminor = 1;
	case DRVOPEN:
	{
		major_t dmajor = cmajor;

		_printd(("%s: driver open\n", __FUNCTION__));
		if (cminor < 1) {
			_printd(("%s: attempt to open minor zero non-clone\n", __FUNCTION__));
			return (ENXIO);
		}
		spin_lock(&nuls_lock);
		for (; *pp && (dmajor = getmajor((*pp)->dev)) < cmajor; pp = &(*pp)->next) ;
		for (; *pp && dmajor == getmajor((*pp)->dev) &&
		     getminor(makedevice(cmajor, cminor)) != 0; pp = &(*pp)->next) {
			minor_t dminor = getminor((*pp)->dev);

			if (cminor < dminor)
				break;
			if (cminor == dminor) {
				if (sflag == CLONEOPEN)
					cminor++;
				else {
					spin_unlock(&nuls_lock);
					kmem_free(p, sizeof(*p));
					pswerr(("%s: stream already open!\n", __FUNCTION__));
					return (EIO);	/* bad error */
				}
			}
		}
		if (getminor(makedevice(cmajor, cminor)) == 0) {	/* no minors left */
			spin_unlock(&nuls_lock);
			kmem_free(p, sizeof(*p));
			_printd(("%s: no minor devices left\n", __FUNCTION__));
			return (EBUSY);	/* no minors left */
		}
		p->dev = *devp = makedevice(cmajor, cminor);
		if ((p->next = *pp))
			p->next->prev = &p->next;
		p->prev = pp;
		*pp = p;
		q->q_ptr = OTHERQ(q)->q_ptr = p;
		spin_unlock(&nuls_lock);
		qprocson(q);
		_printd(("%s: opened major %hu, minor %hu\n", __FUNCTION__, cmajor, cminor));
		return (0);
	}
	}
	pswerr(("%s: bad sflag %d\n", __FUNCTION__, sflag));
	return (ENXIO);
}

static streamscall int
nuls_close(queue_t *q, int oflag, cred_t *crp)
{
	struct nuls *p;

	_trace();
	if ((p = q->q_ptr) == NULL) {
		pswerr(("%s: already closed\n", __FUNCTION__));
		return (0);	/* already closed */
	}
	qprocsoff(q);
	spin_lock(&nuls_lock);
	if ((*(p->prev) = p->next))
		p->next->prev = p->prev;
	p->next = NULL;
	p->prev = &p->next;
	q->q_ptr = OTHERQ(q)->q_ptr = NULL;
	spin_unlock(&nuls_lock);
	_printd(("%s: closed stream with read queue %p\n", __FUNCTION__, q));
	return (0);
}

static struct qinit nuls_rqinit = {
	.qi_putp = NULL,
	.qi_qopen = nuls_open,
	.qi_qclose = nuls_close,
	.qi_minfo = &nuls_minfo,
	.qi_mstat = &nuls_rstat,
};

static struct qinit nuls_wqinit = {
	.qi_putp = nuls_put,
	.qi_srvp = NULL,
	.qi_minfo = &nuls_minfo,
	.qi_mstat = &nuls_wstat,
};

static struct streamtab nuls_info = {
	.st_rdinit = &nuls_rqinit,
	.st_wrinit = &nuls_wqinit,
};

#ifdef LFS
static struct cdevsw nuls_cdev = {
	.d_name = CONFIG_STREAMS_NULS_NAME,
	.d_str = &nuls_info,
	.d_flag = D_MP,
	.d_fop = NULL,
	.d_mode = S_IFCHR | S_IRUGO | S_IWUGO,
	.d_kmod = THIS_MODULE,
};
#endif

#ifdef CONFIG_STREAMS_NULS_MODULE
static
#endif
int __init
nuls_init(void)
{
	int err;

#ifdef CONFIG_STREAMS_NULS_MODULE
	printk(KERN_INFO NULS_BANNER);
#else
	printk(KERN_INFO NULS_SPLASH);
#endif
	nuls_minfo.mi_idnum = modid;
#ifdef LFS
	if ((err = register_strdev(&nuls_cdev, major)) < 0)
		return (err);
#endif
#ifdef LIS
	if ((err = lis_register_strdev(major, &nuls_info, 255, CONFIG_STREAMS_NULS_NAME)) < 0)
		return (err);
#endif
	if (major == 0 && err > 0)
		major = err;
#ifdef LIS
	if ((err = lis_register_driver_qlock_option(major, LIS_QLOCK_NONE)) < 0) {
		lis_unregister_strdev(major);
		return (err);
	}
#endif
	return (0);
};

#ifdef CONFIG_STREAMS_NULS_MODULE
static
#endif
void __exit
nuls_exit(void)
{
#ifdef LFS
	unregister_strdev(&nuls_cdev, major);
#endif
#ifdef LIS
	lis_unregister_strdev(major);
#endif
};

#ifdef CONFIG_STREAMS_NULS_MODULE
module_init(nuls_init);
module_exit(nuls_exit);
#endif
